DM-48013: Add contribution guide and pull request template (#291)
lsst/sdm_schemas: docs/APDB.md

Alert Production Database (APDB) Schema
=======================================

The [apdb](./python/lsst/sdm_schemas/schemas/apdb.yaml) schema describes the Alert Production Database (APDB).
This database is updated by the [Alert Production Pipeline](https://github.com/lsst/ap_pipe) and used for nightly [ap_verify](https://github.com/lsst/ap_verify) runs within continuous integration.

> Previous processing runs may differ from the current schema.

Reviewer: I don't think you need to explain what APDB is and how it is populated in this document.

> The user-queryable Prompt Products Database (PPDB) is expected to have a very similar schema to the APDB.

Reviewer: I don't think we need to say anything about PPDB here, that one is still more or less unknown w.r.t. schema management.

Schema Versioning
-----------------

The versioning system of the APDB is outlined in [DMTN-269](https://dmtn-269.lsst.io/).
The deployed database schema is a product of the [schema in sdm_schemas](yml/apdb.yaml) and processing by tools in the [dax_apdb](https://github.com/lsst/dax_apdb) repository.
Compatibility of the client executable with the actual database schema is determined by the version number in `apdb.yaml` and that which is stored in the database metadata.

> Backward compatibility in the context of the APDB implies that a client using an interface derived from a schema with a higher `MINOR` version number will be able to read and write into a database with a lower `MINOR` version.

Reviewer: Hmm, I don't think I put it like that on the Confluence page. In APDB there is no "interface derived from a schema". There is just schema used by the client to run the queries and an actual schema in the database. It may be better to copy that page more or less verbatim, I think it provides clearer explanation of what compatibility means for APDB?

Reviewer 2: I'm guessing that was a reference to `Apdb`, since I mentioned earlier that we never do direct queries.

Author: I attempted to reword this section so it was clearer and matched your text better.

| Add index | `PATCH` | This should be fully backward compatible.|
| Change object metadata | - | This does not affect the database schema and should not require a version change.|

Client Compatibility
--------------------

Changes to the APDB schema must be propagated to the API defined in the [api_packet](https://github.com/lsst/alert_packet) repository, following the instructions under [Adding a new schema](https://github.com/lsst/alert_packet/tree/main?tab=readme-ov-file#adding-a-new-schema).

Database Migrations
-------------------

The APDB database schema is upgraded using a special migration tool defined in the [dax_apdb_migrate](https://github.com/lsst-dm/dax_apdb_migrate) repository. For every version change to `apdb.yaml`, a new migration script needs to be added to that repository. The details of the migration process are described in the [package documentation](https://pipelines.lsst.io/modules/lsst.dax.apdb/index.html). Migration scripts should normally be implemented on the same Jira ticket as the corresponding schema change.

Reviewers
---------

> - [Andy Salnikov](https://github.com/andy-slac)

Reviewer: I should not be the only reviewer for APDB schema changes, but I can be added to review a corresponding version change.

Author: This is more like who to select as a default if it is otherwise unknown.

Reviewer: This is still not right, I would know even less than PR author whom to call. Maybe add @isullivan instead or whomever Ian recommends.

Alert Production Database (APDB) Schema
=======================================

> The [apdb](../python/lsst/sdm_schemas/schemas/apdb.yaml) schema describes the Alert Production Database (APDB),
> which contains the results of processing from the [Alert Production Pipeline](https://github.com/lsst/ap_pipe).

Reviewer: ```suggestion
The [apdb](../python/lsst/sdm_schemas/schemas/apdb.yaml) schema describes the Alert Production Database (APDB), which contains the results of processing from the [Alert Production Pipeline](https://github.com/lsst/ap_pipe).
```

Schema Versioning
-----------------

The versioning system of the APDB is outlined in [DMTN-269](https://dmtn-269.lsst.io/).

> The deployed database schema is a product of the [schema in sdm_schemas](../python/lsst/sdm_schemas/schemas/apdb.yaml) and processing by tools in the [dax_apdb](https://github.com/lsst/dax_apdb) repository.

Reviewer: ```suggestion
The deployed database schema is a product of the [schema in sdm_schemas](../python/lsst/sdm_schemas/schemas/apdb.yaml) with additional processing by APDB client library in the [dax_apdb](https://github.com/lsst/dax_apdb) repository.
```

There may be cases where significant schema changes may still be compatible if the client can handle them transparently.

Typical Schema Changes
----------------------

Here are some guidelines for incrementing the version number based on what changes are being performed:

| Operation | Increment | Notes |
|--------|-----------|-------|
| Add table | `MAJOR` | The client using the new schema will typically want to read or write into that table, which will make it incompatible.|
| Add column | `MAJOR` | The client using the new schema will typically want to read or write into that column, which will make it incompatible.|
| Remove column | `MINOR` or `PATCH` | Clients using the new schema will not care about a removed column, so this should be backward compatible.|
| Change column type | `MINOR` or `PATCH` | This may or may not be completely transparent and so could be either backward compatible or fully compatible.
| Add index | `PATCH` | This should be fully backward compatible.|
| Change object metadata | - | This does not affect the database schema and should not require a version change.|

Client Compatibility
--------------------

> Changes to the APDB schema must be propagated to the API defined in the [api_packet](https://github.com/lsst/alert_packet) repository, following the instructions under [Adding a new schema](https://github.com/lsst/alert_packet/tree/main?tab=readme-ov-file#adding-a-new-schema).

Reviewer 3: ```suggestion
Changes to the APDB schema must be propagated to the Avro alert schemas defined in the [alert_packet](https://github.com/lsst/alert_packet) repository, following the instructions under [Adding a new schema](https://github.com/lsst/alert_packet/tree/main?tab=readme-ov-file#adding-a-new-schema).
```

Database Migrations
-------------------

The APDB database schema is upgraded using a special migration tool defined in the [dax_apdb_migrate](https://github.com/lsst-dm/dax_apdb_migrate) repository.
For every version change to `apdb.yaml`, a new migration script needs to be added to that repository.

> The details of the migration process are described in the [package documentation](https://pipelines.lsst.io/modules/lsst.dax.apdb/index.html).

Reviewer: The link is incorrect. We do not have published docs for migration script yet, but better link should be:
```suggestion
The details of the migration process are described in the [package documentation](https://github.com/lsst-dm/dax_apdb_migrate/blob/main/doc/lsst.dax.apdb_migrate/index.rst).
```

> Migration scripts should normally be implemented on the same Jira ticket as the corresponding schema change.

Reviewer 2: @andy-slac, should this mention the distinction between SQL and Cassandra (or the fact that the latter is currently manual rather than scripted)?

Author: > the distinction between SQL and Cassandra

SQL = Postgres?

Reviewer: I do not want to mention manual part here, it's mostly irrelevant, but mentioning Cassandra is probably worth it. @JeremyMcCormick, could you add a sentence here:
```
APDB has a Cassandra-based implementation in addition to a SQL-based, and Cassandra may impose additional restrictions on possible schema changes.
```
